lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,460 @@
+	sessionCtx *common.Session
+}
+
+// InitializeConnection initializes the database connection.
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.proxyConn = clientConn
+	e.clientReader = redis.NewReader(clientConn)
+	e.sessionCtx = sessionCtx
+
+	return nil
+}
+
+// authorizeConnection does authorization check for MongoDB connection about
+// to be established.
+func (e *Engine) authorizeConnection(ctx context.Context, sessionCtx *common.Session) error {
+	ap, err := e.Auth.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	mfaParams := services.AccessMFAParams{
+		Verified:       sessionCtx.Identity.MFAVerified != "",
+		AlwaysRequired: ap.GetRequireSessionMFA(),
+	}
+
+	dbRoleMatchers := role.DatabaseRoleMatchers(
+		sessionCtx.Database.GetProtocol(),
+		sessionCtx.DatabaseUser,
+		// Leave database empty as Redis integration doesn't support schema access control.
+		"",
+	)
+	err = sessionCtx.Checker.CheckAccess(
+		sessionCtx.Database,
+		mfaParams,
+		dbRoleMatchers...,
+	)
+	if err != nil {
+		e.Audit.OnSessionStart(e.Context, sessionCtx, err)
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// SendError sends error message to connected client.
+func (e *Engine) SendError(redisErr error) {
+	if redisErr == nil || utils.IsOKNetworkError(redisErr) {
+		return
+	}
+
+	if err := e.sendToClient(redisErr); err != nil {
+		e.Log.Errorf("failed to send message to the client: %v", err)
+		return
+	}
+}
+
+// sendToClient sends a command to connected Redis client.
+func (e *Engine) sendToClient(vals interface{}) error {
+	buf := &bytes.Buffer{}
+	wr := redis.NewWriter(buf)
+
+	if err := writeCmd(wr, vals); err != nil {
+		return trace.BadParameter("failed to convert error to a message: %v", err)
+	}
+
+	if _, err := e.proxyConn.Write(buf.Bytes()); err != nil {
+		return trace.ConnectionProblem(err, "failed to send message to the client")
+	}
+
+	return nil
+}
+
+// HandleConnection is responsible for connecting to a Redis instance/cluster and
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session) error {
+	// Check that the user has access to the database.
+	err := e.authorizeConnection(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err, "error authorized database access")
+	}
+
+	tlsConfig, err := e.Auth.GetTLSConfig(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	connectionOptions, err := ParseRedisURI(sessionCtx.Database.GetURI())
+	if err != nil {
+		return trace.BadParameter("Redis connection string is incorrect: %v", err)
+	}
+
+	var (
+		redisConn      redis.UniversalClient
+		connectionAddr = fmt.Sprintf("%s:%s", connectionOptions.address, connectionOptions.port)
+	)
+
+	// TODO(jakub): Use system CA bundle if connecting to AWS.
+	// TODO(jakub): Investigate Redis Sentinel.
+	if connectionOptions.cluster {
+		redisConn = redis.NewClusterClient(&redis.ClusterOptions{
+			Addrs:     []string{connectionAddr},
```

Comment:
for sharded redis cluster, shouldn't we actually pass a list of all servers here?

Reply:
We could, but it's not required. It's enough to pass only one server, then `go-redis` will get the list of all available masters from that server. I think that as some point we may add that option, but to keep it simple for now I decided to stick with only one.